In [26]:
!pip install mlflow

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 29.0/29.0 MB 16.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.3/6.3 MB 68.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.9/114.9 kB 6.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 85.0/85.0 kB 5.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 733.7/733.7 kB 29.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 203.4/203.4 kB 13.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 118.5/118.5 kB 8.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 196.2/196.2 kB 13.3 MB/s eta 0:00:00


In [27]:
import pandas as pd
import numpy as np
import pickle
import mlflow
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.feature_extraction import DictVectorizer
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge

from sklearn.metrics import mean_squared_error
from sklearn.metrics import root_mean_squared_error

### Q1. Select the Tool

In [2]:
!pip install prefect

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.7/53.7 kB 2.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 49.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 242.5/242.5 kB 16.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 51.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 66.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.2/98.2 kB 7.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 295.7/295.7 kB 13.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 147.8/147.8 kB 9.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.3/129.3 kB 7.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.8/65.8 kB 3.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 353.7/353.7 kB 20.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.4/44.4 kB 2.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.0/55.

Answer: I have used **Prefect** for orchestration

### Q2. Version

In [4]:
!prefect version

Version:             3.4.5
API version:         0.8.4
Python version:      3.11.13
Git commit:          df37c8cf
Built:               Sat, Jun 07, 2025 02:24 AM
OS/Arch:             linux/x86_64
Profile:             ephemeral
Server type:         ephemeral
Pydantic version:    2.11.5
Server:
  Database:          sqlite
  SQLite version:    3.37.2


Answer: The version of Prefect is **3.4.5**

### Q3. Creating a pipeline

In [12]:
df = pd.read_parquet('data/yellow_tripdata_2023-03.parquet')
df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-03-01 00:06:43   2023-03-01 00:16:43              1.0   
1         2  2023-03-01 00:08:25   2023-03-01 00:39:30              2.0   
2         1  2023-03-01 00:15:04   2023-03-01 00:29:26              0.0   
3         1  2023-03-01 00:49:37   2023-03-01 01:01:05              1.0   
4         2  2023-03-01 00:08:04   2023-03-01 00:11:06              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           0.00         1.0                  N           238            42   
1          12.40         1.0                  N           138           231   
2           3.30         1.0                  N           140           186   
3           2.90         1.0                  N           140            43   
4           1.23         1.0                  N            79           137   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          8.6    1.0      0.5        0.00           0.0   
1             1         52.7    6.0      0.5       12.54           0.0   
2             1         18.4    3.5      0.5        4.65           0.0   
3             1         15.6    3.5      0.5        4.10           0.0   
4             1          7.2    1.0      0.5        2.44           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  Airport_fee  
0                    1.0         11.10                   0.0         0.00  
1                    1.0         76.49                   2.5         1.25  
2                    1.0         28.05                   2.5         0.00  
3                    1.0         24.70                   2.5         0.00  
4                    1.0         14.64                   2.5         0.00

In [13]:
df.shape

(3403766, 19)

Answer: There are **3403766** records in the dataset.

### Q4. Data preparation

In [14]:
df['duration'] = df.tpep_dropoff_datetime - df.tpep_pickup_datetime
df.duration = df.duration.dt.total_seconds() / 60

df = df[(df.duration >= 1) & (df.duration <= 60)]

categorical = ['PULocationID', 'DOLocationID']
df[categorical] = df[categorical].astype(str)

df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-03-01 00:06:43   2023-03-01 00:16:43              1.0   
1         2  2023-03-01 00:08:25   2023-03-01 00:39:30              2.0   
2         1  2023-03-01 00:15:04   2023-03-01 00:29:26              0.0   
3         1  2023-03-01 00:49:37   2023-03-01 01:01:05              1.0   
4         2  2023-03-01 00:08:04   2023-03-01 00:11:06              1.0   

   trip_distance  RatecodeID store_and_fwd_flag PULocationID DOLocationID  \
0           0.00         1.0                  N          238           42   
1          12.40         1.0                  N          138          231   
2           3.30         1.0                  N          140          186   
3           2.90         1.0                  N          140           43   
4           1.23         1.0                  N           79          137   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          8.6    1.0      0.5        0.00           0.0   
1             1         52.7    6.0      0.5       12.54           0.0   
2             1         18.4    3.5      0.5        4.65           0.0   
3             1         15.6    3.5      0.5        4.10           0.0   
4             1          7.2    1.0      0.5        2.44           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  Airport_fee  \
0                    1.0         11.10                   0.0         0.00   
1                    1.0         76.49                   2.5         1.25   
2                    1.0         28.05                   2.5         0.00   
3                    1.0         24.70                   2.5         0.00   
4                    1.0         14.64                   2.5         0.00   

    duration  
0  10.000000  
1  31.083333  
2  14.366667  
3  11.466667  
4   3.033333

In [15]:
df.shape

(3316216, 20)

Answer: There are **3316216** rows after preprocessing

### Q5. Train a model

In [16]:
df_copy1 = df.copy()

In [17]:
dv = DictVectorizer()
train_dicts = df_copy1[categorical].to_dict(orient='records')
X_train = dv.fit_transform(train_dicts)

target = 'duration'
y_train = df_copy1[target].values

lr = LinearRegression()
lr.fit(X_train, y_train)

LinearRegression()

In [19]:
lr.intercept_

np.float64(24.77203445209766)

Answer: The intercept of the model is **24.77**

### Q6. Register the model

In [29]:
!mlflow ui --backend-store-uri sqlite:///mlflow.db --default-artifact-root ./artifacts

2025/06/10 10:38:11 INFO mlflow.store.db.utils: Creating initial MLflow database tables...
2025/06/10 10:38:11 INFO mlflow.store.db.utils: Updating database tables
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade  -> 451aebb31d03, add metric step
INFO  [alembic.runtime.migration] Running upgrade 451aebb31d03 -> 90e64c465722, migrate user column to tags
INFO  [alembic.runtime.migration] Running upgrade 90e64c465722 -> 181f10493468, allow nulls for metric values
INFO  [alembic.runtime.migration] Running upgrade 181f10493468 -> df50e92ffc5e, Add Experiment Tags Table
INFO  [alembic.runtime.migration] Running upgrade df50e92ffc5e -> 7ac759974ad8, Update run tags with larger limit
INFO  [alembic.runtime.migration] Running upgrade 7ac759974ad8 -> 89d4b8295536, create latest metrics table
INFO  [89d4b8295536_create_latest_metrics_table_py] Migration complete!
INFO  

In [30]:
mlflow.set_tracking_uri("sqlite:///mlflow.db")
mlflow.set_experiment("nyc_taxi_experiment")

with mlflow.start_run():
    mlflow.set_tag("developer", "Krishna")
    mlflow.set_tag("model", "LinearRegression")
    lr = LinearRegression()
    lr.fit(X_train, y_train)

    mlflow.log_param("intercept_", lr.intercept_)
    mlflow.sklearn.log_model(
        sk_model=lr,
        artifact_path="artifacts_local",
        registered_model_name="MyLinearRegressor"
    )

2025/06/10 10:39:38 INFO mlflow.tracking.fluent: Experiment with name 'nyc_taxi_experiment' does not exist. Creating a new experiment.
2025/06/10 10:41:05 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
Successfully registered model 'MyLinearRegressor'.
Created version '1' of model 'MyLinearRegressor'.


In [36]:
!cat mlruns/1/cf2f88b411dd4704984fe636e1a67c32/artifacts/artifacts_local/MLmodel

artifact_path: artifacts_local
flavors:
  python_function:
    env:
      conda: conda.yaml
      virtualenv: python_env.yaml
    loader_module: mlflow.sklearn
    model_path: model.pkl
    predict_fn: predict
    python_version: 3.11.13
  sklearn:
    code: null
    pickled_model: model.pkl
    serialization_format: cloudpickle
    sklearn_version: 1.6.1
mlflow_version: 2.22.1
model_size_bytes: 4501
model_uuid: 9b4e02dc82034ae0b7402a5f8eb7b672
prompts: null
run_id: cf2f88b411dd4704984fe636e1a67c32
utc_time_created: '2025-06-10 10:40:53.313379'


Answer: The model size is **4501**, nearest match is **4534**